src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs

```diff
@@ -264,15 +265,38 @@ private void CheckNotDisposed()
         /// In the case of IBufferWriter, this advances the underlying <see cref="IBufferWriter{Byte}" /> based on what has been written so far.
         /// In the case of Stream, this writes the data to the stream and flushes it.
         /// </remarks>
+        /// <exception cref="ObjectDisposedException">
+        ///   The instance of <see cref="Utf8JsonWriter"/> has been disposed.
+        /// </exception>
         public void Flush()
         {
-            FlushHelper();
+            CheckNotDisposed();
 
             if (_stream != null)
             {
-                FlushHelperStream();
+                Debug.Assert(_arrayBufferWriter != null);
+                if (BytesPending != 0)
+                {
+                    _arrayBufferWriter.Advance(BytesPending);
+                    Debug.Assert(BytesPending == _arrayBufferWriter.WrittenCount);
+#if BUILDING_INBOX_LIBRARY
+                    _stream.Write(_arrayBufferWriter.WrittenSpan);
+#else
+                    Debug.Assert(_arrayBufferWriter.WrittenSpan.Length == _arrayBufferWriter.WrittenCount);
+                    _stream.Write(_arrayBufferWriter.WrittenSpan.ToArray(), 0, _arrayBufferWriter.WrittenCount);
```

Comment:
How much do we care about optimizing the source package?  Wondering if this should rent a pool array, copy to that, write that array, then return the array to the pool.

Reply:
Or, actually, we know the ArrayBufferWriter wraps an array, so _arrayBufferWriter.WrittenMemory.TryGetArray should let you retrieve the underlying array to write, right?

Reply:
Good point, we can do that.

Reply:
@steveharter, we should probably do something similar here: 
https://github.com/dotnet/corefx/blob/1f9b84a0804e868c7e0f37a3c10fbaf7c735ae14/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Stream.cs#L63-L68
https://github.com/dotnet/corefx/blob/1f9b84a0804e868c7e0f37a3c10fbaf7c735ae14/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Stream.cs#L91-L96